### ISSUE TITLE
Support a kopia-based backup

### ISSUE THREAD

**Issue by onedr0p** [backube/volsync#320]
**Describe the feature you'd like to have.**

Hi again 👋🏼 

Would it be possible to support Kopia? I know you already have support for restic, rclone and rsync, but I figured Kopia would fit in nicely as it's becoming a very popular tool for incremental data backup. In my tests, it is also very fast.

https://github.com/kopia/kopia

It is written in Go and has decent APIs

https://pkg.go.dev/github.com/kopia/kopia/repo?utm_source=godoc

**What is the value to the end user? (why is it a priority?)**

Supporting another way of backing and restoring data.

**How will we know we have a good solution? (acceptance criteria)**

**Additional context**

**Comment by onedr0p** [comment]
Yeah those benchmarks are pretty clear that Kopia is a serious contender to restic :)

**Comment by trieded** [comment]
hi any plans for this?

**Comment by onedr0p** [comment]
There's needs to be some considerations made on how to implement this since kopia works a little different than existing movers. For example kopia can support a single "kopia repository" for multiple apps and "kopia snapshots" are the way you backup or restore data.

**Comment by onedr0p** [comment]
#  Local Filesystem Commands

This method is for using Kopia on your filesystem, not backing up to s3. Kopia does support s3 but I need to look into how that works. It's probably quite similar.

- `/snapshots` is the location of the Kopia Repository
- `/data` is the location of the PVC root filesystem

Note that in the examples below `/snapshots` is a single Kopia Repository with multiple snapshots across all PVCs that are backed up. This can help with deduplication if all the PVC volumes are in the same Kopia Repository. It could also be architected to support a 1:1 snapshot to Kopia Repository like the restic implementation is currently doing.

At a glance in the Kopia web UI, a single Kopia Repository with multiple snapshots (PVCs) looks like this:

<img width="1289" alt="image" src="https://user-images.githubusercontent.com/213795/204154483-cf38f274-e343-4770-b942-61130e864466.png">

## Create backup

```bash
[[ ! -f /snapshots/kopia.repository.f ]] && kopia repository create filesystem --path=/snapshots
kopia repo connect filesystem --path=/snapshots --override-hostname=cluster --override-username=root
kopia policy set /data/<namespace/<app>/<claim> --compression=zstd --keep-latest 14 --keep-hourly 0 --keep-daily 7 --keep-weekly 2 --keep-monthly 0 --keep-annual 0
kopia snap create /data/<namespace/<app>/<claim>
kopia repo disconnect
```

## Restore Backup

```bash
# Get latest snapshot ID
SNAPSHOT=$(kopia -- kopia snapshot list /data/<namespace/<app>/<claim> --json | jq --raw-output '.[-1] | .id')
```

```bash
kopia repo connect filesystem --path=/snapshots --override-hostname=cluster --override-username=root
kopia snap restore "${SNAPSHOT}" /data/<namespace/<app>/<claim>
kopia repo disconnect
```

## Maintenance 

```bash
kopia repo connect filesystem --path=/snapshots --override-hostname=cluster --override-username=root
kopia maintenance set --owner=root@cluster
kopia maintenance run --full
kopia repo disconnect
```

I hope this can help lend some help on implementing this feature into VolSync.

**Comment by onedr0p** [comment]
I am sure you are all open to PRs @JohnStrunk to add Kopia support?

**Comment by perfectra1n** [comment]
I'd be happy to help with the PR as well, if John is open to PRs that add Kopia's support.

**Comment by JohnStrunk** [comment]
@tesshuflower and I discussed this a bit, and we'd like to find a way that we can take more community contributions to the project. So, if you all are interested in contributing a kopia mover, we're happy to shepherd it through.

By the end of the process, we need to have:
- Mover code (controller + mover parts)
- Documentation
- Tests

Since this will be a fairly substantial piece, starting w/ a design PR is probably a good place to begin.

**Comment by DreamingRaven** [comment]
Hey guys, just dropping a ping here.

Any news regarding this? I would also be happy to lend a hand with this.

Kopia theoretically seems more appropriate for volumes with a plethora of small files, which I tend to find seems to be the predominant scenario in my volumes. https://cloudcasa.io/blog/comparing-restic-vs-kopia-for-kubernetes-data-movement/

**Comment by onedr0p** [comment]
There is a PR opened for this #1723 Hopefully we can get it reviewed. I and a few others are kicking the tires on it.

**Comment by perfectra1n** [comment]
> [@tesshuflower](https://github.com/tesshuflower) and I discussed this a bit, and we'd like to find a way that we can take more community contributions to the project. So, if you all are interested in contributing a kopia mover, we're happy to shepherd it through.
> 
> By the end of the process, we need to have:
> 
> * Mover code (controller + mover parts)
> * Documentation
> * Tests
> 
> Since this will be a fairly substantial piece, starting w/ a design PR is probably a good place to begin.

So I kinda skipped the design PR and moved to the "end of process" steps

**Comment by onedr0p** [comment]
Your PR description and docs are a pretty good design implementation, so hopefully that can work for that step.

**Comment by onedr0p** [comment]
I am closing this issue as I moved onto another tool suited more to my needs: https://github.com/home-operations/kopiur/

Thanks to all the volsync contributors during my time using it, and if you feel like this is something you still want to tackle please don't hesitate to re-open.
